File: src/types/specialAbility/MagicalSign.ts
import * as DB from "tsondb/schema/dsl"
import { effect, name, name_in_library } from "../_Activatable.js"
import { ap_value_append, ap_value_l10n } from "../_ActivatableAdventurePointsValue.js"
import { propertyOptional } from "../_ActivatableNonMundane.js"
import { OneTimeCostMap } from "../_ActivatableSkillCost.js"
import { GeneralPrerequisites } from "../_Prerequisite.js"
import { NestedTranslationMap } from "../Locale.js"
import { Errata } from "../source/_Erratum.js"
import { src } from "../source/_PublicationRef.js"

export const MagicalSign = DB.Entity(import.meta.url, {
  name: "MagicalSign",
  namePlural: "MagicalSigns",
  type: () =>
    DB.Object({
      prerequisites: DB.Optional({
        type: DB.IncludeIdentifier(GeneralPrerequisites),
      }),
      cost: DB.Optional({
        comment: "The cost in AE.",
        type: DB.IncludeIdentifier(MagicalSignCost),
      }),
      property: propertyOptional(),
      ap_value: DB.Required({
        comment: "The adventure points value.",
        type: DB.Integer({ minimum: 0 }),
      }),
      src,
      translations: NestedTranslationMap(
        DB.Required,
        "MagicalSign",
        DB.Object({
          name,
          name_in_library,
          effect,
          ap_value_append,
          ap_value: ap_value_l10n,
          errata: DB.Optional({
            type: DB.IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  instanceDisplayName: {},
  uniqueConstraints: [
    {
      entityMapKeyPath: "translations",
      keyPathInEntityMap: "name_in_library",
      keyPathInEntityMapFallback: "name",
    },
  ],
})

const MagicalSignCost = DB.Enum(import.meta.url, {
  name: "MagicalSignCost",
  values: () => ({
    Constant: DB.EnumCase({ type: DB.IncludeIdentifier(ConstantMagicalSignCost) }),
    Map: DB.EnumCase({ type: DB.IncludeIdentifier(OneTimeCostMap) }),
  }),
})

const ConstantMagicalSignCost = DB.TypeAlias(import.meta.url, {
  name: "ConstantMagicalSignCost",
  type: () =>
    DB.Object({
      value: DB.Required({
        comment: "The AE cost value.",
        type: DB.Integer({ minimum: 1 }),
      }),
      permanent_value: DB.Optional({
        comment: "The part of the cost value that has to be spent permanently.",
        type: DB.Integer({ minimum: 1 }),
      }),
    }),
})

File: Sources/OptolithDatabaseSchema/GeneratedTypes/specialAbility/MagicalSign.swift
//
//  MagicalSign.swift
//  OptolithDatabaseSchema
//

import DiscriminatedEnum

public struct MagicalSign: LocalizableEntity {
    public let id: Id
    
    public let prerequisites: GeneralPrerequisites?
    
    /// The associated magic property.
    public let property: PropertyReference
    
    /// The adventure points value.
    public let apValue: Int
    
    public let src: PublicationRefs
    
    /// All translations for the entry, identified by IETF language tag (BCP47).
    public let translations: LocaleMap<MagicalSignTranslation>

    public init(id: Id, prerequisites: GeneralPrerequisites? = nil, property: PropertyReference, apValue: Int, src: PublicationRefs, translations: LocaleMap<MagicalSignTranslation>) {
        self.id = id
        self.prerequisites = prerequisites
        self.property = property
        self.apValue = apValue
        self.src = src
        self.translations = translations
    }    
    
    private enum CodingKeys: String, CodingKey {
        case id = "id"
        case prerequisites = "prerequisites"
        case property = "property"
        case apValue = "ap_value"
        case src = "src"
        case translations = "translations"
    }
}

public struct MagicalSignTranslation: EntitySubtype {
    public let name: Name
    
    public let nameInLibrary: NameInLibrary?
    
    public let effect: Effect
    
    public let errata: Errata?

    public init(name: Name, nameInLibrary: NameInLibrary? = nil, effect: Effect, errata: Errata? = nil) {
        self.name = name
        self.nameInLibrary = nameInLibrary
        self.effect = effect
        self.errata = errata
    }    
    
    private enum CodingKeys: String, CodingKey {
        case name = "name"
        case nameInLibrary = "name_in_library"
        case effect = "effect"
        case errata = "errata"
    }
}
